src/librustc/ty/context.rs

```diff
@@ -1658,59 +1687,120 @@ pub mod tls {
         })
     }
 
-    pub fn enter_global<'gcx, F, R>(gcx: GlobalCtxt<'gcx>, f: F) -> R
-        where F: for<'a> FnOnce(TyCtxt<'a, 'gcx, 'gcx>) -> R
+    fn track_diagnostic(diagnostic: &Diagnostic) {
+        with_context(|context| {
+            if let Some(ref query) = context.query {
+                query.diagnostics.lock().push(diagnostic.clone());
+            }
+        })
+    }
+
+    pub fn with_thread_locals<F, R>(f: F) -> R
```

Comment:
Comment please. (I know... the original code isn't that documented either...c'est la vie)